### ISSUE TITLE
GraphMultisetTransformer does not converge

### ISSUE THREAD

**Issue by mazabou** [pyg-team/pytorch_geometric#3482]
Thank you @JinheonBaek for the GMT implementation! 
Unfortunately, the `GraphMultisetTransformer` pooling layer is not working correctly for me. The loss does not converge for some reason. This happens for me even when I run the example code in `examples/proteins_gmt.py`:

```
...
Epoch: 167, Loss: 0.4678, Val Acc: 0.6786, Test Acc: 0.7321
Epoch: 168, Loss: 0.4652, Val Acc: 0.6786, Test Acc: 0.7054
Epoch: 169, Loss: 0.4565, Val Acc: 0.6786, Test Acc: 0.7321
Epoch: 170, Loss: 0.4542, Val Acc: 0.6696, Test Acc: 0.7143
Epoch: 171, Loss: 0.4424, Val Acc: 0.6875, Test Acc: 0.7679
Epoch: 172, Loss: 0.4492, Val Acc: 0.6696, Test Acc: 0.7321
Epoch: 173, Loss: 0.4453, Val Acc: 0.6786, Test Acc: 0.7321
Epoch: 174, Loss: 0.4523, Val Acc: 0.6875, Test Acc: 0.7589
Epoch: 175, Loss: 0.4521, Val Acc: 0.6786, Test Acc: 0.7321
Epoch: 176, Loss: 0.4531, Val Acc: 0.6875, Test Acc: 0.7232
Epoch: 177, Loss: 0.4679, Val Acc: 0.6786, Test Acc: 0.7143
Epoch: 178, Loss: 0.4636, Val Acc: 0.6786, Test Acc: 0.7589
Epoch: 179, Loss: 0.4695, Val Acc: 0.6786, Test Acc: 0.7589
Epoch: 180, Loss: 0.4686, Val Acc: 0.6786, Test Acc: 0.7500
Epoch: 181, Loss: 0.4678, Val Acc: 0.6786, Test Acc: 0.7143
Epoch: 182, Loss: 0.4531, Val Acc: 0.6786, Test Acc: 0.7411
Epoch: 183, Loss: 0.4693, Val Acc: 0.6875, Test Acc: 0.7411
Epoch: 184, Loss: 0.4622, Val Acc: 0.6786, Test Acc: 0.7411
Epoch: 185, Loss: 0.4639, Val Acc: 0.6607, Test Acc: 0.7321
Epoch: 186, Loss: 0.4529, Val Acc: 0.6786, Test Acc: 0.7500
Epoch: 187, Loss: 0.4384, Val Acc: 0.6786, Test Acc: 0.7321
Epoch: 188, Loss: 0.4579, Val Acc: 0.6875, Test Acc: 0.7143
Epoch: 189, Loss: 0.4608, Val Acc: 0.6964, Test Acc: 0.7321
Epoch: 190, Loss: 0.4412, Val Acc: 0.6786, Test Acc: 0.7500
Epoch: 191, Loss: 0.4548, Val Acc: 0.6964, Test Acc: 0.7143
```

Any idea why this is a problem?

All other components from PyTorch Geometric work as expected. This is my environment:
* OS: Ubuntu 20.04
* Python version: 3.8.10
* CUDA/cuDNN version: 11.2
* GCC version: 9.3.0

Package versions:
torch==1.9.0+cu111
torch-cluster==1.5.9
torch-geometric==2.0.2
torch-scatter==2.0.9
torch-sparse==0.6.12
torch-spline-conv==1.2.1

**Comment by rusty1s** [comment]
Does this only happen on `PROTEINS` or other datasets as well? The loss is decreasing for me, although accuracy fluctuates quite a bit (something which is quite normal on `TUDataset`s though).

**Comment by mazabou** [comment]
It happens for my dataset (a set classification problem) for which I was using the `GlobalAttention` pooling head with no problem. 

then when I use this:
```
pool = GraphMultisetTransformer(
            in_channels=32,
            hidden_channels=32,
            out_channels=32,
            num_nodes=400,
            pooling_ratio=0.1,
            pool_sequences=["GMPool_I", "SelfAtt", "GMPool_I"],  # also tried ["GMPool_I"] 
            num_heads=4,
            layer_norm=True,  # also tried False 
        )
```

the loss is stagnant and noisy even when i change learning rate.

so I tried the PROTEINS example, to make sure that it wasn't a problem with my setup and found a similar fluctuation. The loss does slightly decrease at the beginning but I was not sure whether it was simply due to training of the final classification layers.

I can try to run this on a different standard dataset if you have any suggestions.

**Comment by rusty1s** [comment]
@JinheonBaek Shall we look into the PyG `GMT` implementation once again to see if it is correctly implemented?

**Comment by JinheonBaek** [comment]
Sure, I will check as soon as possible, maybe within one week.

The below are the few points that might cause the problem of not convergence.

---

Note that, the attention operation generally requires a large number of training data. Thus, if you use your own dataset having a small number of nodes, then it might cause a problem. In other words, I think you might have lesser problems on large-scale datasets having a large number of nodes, such as D&D.

Also, one other possible reason is that the performance of models on small-scale graph classification datasets is significantly different across different folds of training data [1] especially on the TU dataset, thus you might fail into an unlucky case where the training fold is quite different from the test fold. One way to handle this is to use the publicly available splits of the graph classification datasets from TU [2] or OGB [3].

[1] Errica et al. A Fair Comparison of Graph Neural Networks for Graph Classification. ICLR 2020.
[2] https://github.com/JinheonBaek/GMT/tree/main/datasets
[3] Hu et al. Open Graph Benchmark: Datasets for Machine Learning on Graphs. NeurIPS 2020.

**Comment by JinheonBaek** [comment]
Thank you @mazabou for your interest in our work.

It seems the training loss of our GMT is decreasing for me, however, it somewhat fluctuates compared to other methods (e.g., MinCutPool) on the proteins dataset, which might be due to the attention operation that has more number of parameters unlike simple MLPs or GNNs. However, the accuracy is indeed increasing.

Specifically, I test the GMT and MinCutPool on the proteins dataset, where I strictly set the random seed as 42, and the number of training epochs as 1,000. In the training log, the GMT starts with a loss of 0.9222, and ends with a loss of 0.3963. Similarly, the MinCutPool starts with a loss of 0.644, and ends with a loss of 0.373. In this vein, I believe there are no problems on the implementation of GMT.

I further checked that the parameters of GMT are updated (changed) during training. You can check this by printing out the weights of our model parameters (e.g., model.pool.pools[0].mab.fc_q.weight).
